Src/Support/Google.Apis.Auth/OAuth2/ImpersonatedCredential.cs

```diff
@@ -0,0 +1,264 @@
+                IEnumerable<string> Delegates = null, IEnumerable<string> Scopes = null, int LifetimeInSeconds = 3600)
+                : base(String.Format(GoogleAuthConsts.IamAccessTokenEndpoint, TargetPrincipal)) { 
+                    this.SourceCredential = SourceCredential;
+                    this.TargetPrincipal = TargetPrincipal;
+                    this.Delegates = Delegates;
+                    this.Scopes = Scopes;
+                    this.LifetimeInSeconds = LifetimeInSeconds;
+                }
+
+            internal Initializer(ImpersonatedCredential other) : base(other.TokenServerUrl)
+            {
+                this.SourceCredential = other.SourceCredential;
+                this.TargetPrincipal = other.TargetPrincipal;
+                this.Delegates = other.Delegates;
+                this.Scopes = other.Scopes;
+                this.LifetimeInSeconds = other.LifetimeInSeconds;
+            }
+        }
+
+        /// <summary>
+        /// Gets the source credential used as to acquire the impersonated credentials.
+        /// </summary>
+        public GoogleCredential SourceCredential { get; }
+
+        /// <summary>
+        /// Gets the service account to impersonate.
+        /// </summary>
+        public string TargetPrincipal { get; }
+
+        /// <summary>
+        /// Gets the chained list of delegates.
+        /// </summary>
+        public IEnumerable<string> Delegates { get; }
+
+        /// <summary>
+        /// Gets the scopes to request during the authorization grant.
+        /// </summary>
+        public IEnumerable<string> Scopes { get; }
+
+        /// <summary>
+        /// Gets the number of seconds the delegated credential should be valid.
+        /// </summary>
+        public int LifetimeInSeconds { get; }
+
+        /// <summary>Constructs a new impersonated credential using the given initializer.</summary>
+        internal ImpersonatedCredential(Initializer initializer) : base(initializer)
+        {
+            SourceCredential = initializer.SourceCredential.ThrowIfNull("initializer.SourceCredential");
+            if (!(SourceCredential.UnderlyingCredential is UserCredential || SourceCredential.UnderlyingCredential is ServiceAccountCredential))
+            {
+                throw new InvalidOperationException("The underlying credential of source credential must be UserCredential or ServiceAccountCredential.");
+            }
+            SourceCredential = SourceCredential.CreateScoped(new string[] {GoogleAuthConsts.IamScope});
+            HttpClient.MessageHandler.Credential = SourceCredential.UnderlyingCredential as Http.IHttpExecuteInterceptor;
+            
+            TargetPrincipal = initializer.TargetPrincipal.ThrowIfNullOrEmpty("initializer.TargetPrincipal");
+            Delegates = initializer.Delegates;
+            Scopes = initializer.Scopes;
+            
+            if (initializer.LifetimeInSeconds > 43200)
```

Comment:
As noted above, I think it would be better to do defaulting and validation in the initializer constructor.

Reply:
done